pkg/stream/stream.go

```diff
@@ -0,0 +1,160 @@
+	rep, err := OpenKopiaRepository(ctx, password)
+	if err != nil {
+		return errors.Wrap(err, "Failed to open kopia repository")
+	}
+	// Initialize a directory tree with given file
+	// The following will create <dirPath>/<filePath> objects
+	// Example: If dirPath is `/mnt/data` and filePath is `dir/file`,
+	// `data` will be the root directory and
+	// `dir/file` objects will be created under it
+	root, err := virtualfs.NewDirectory(filepath.Base(dirPath))
+	if err != nil {
+		return errors.Wrap(err, "Failed to create root directory")
+	}
+	if _, err = virtualfs.AddFileWithStreamSource(root, filePath, sourceEndpoint, defaultPermissions, defaultPermissions); err != nil {
+		return errors.Wrap(err, "Failed to add file with the given stream source to the root directory")
+	}
+
+	// Setup kopia uploader
+	u := snapshotfs.NewUploader(rep)
+
+	// Populate the source info with source path and file
+	sourceInfo := snapshot.SourceInfo{
+		UserName: rep.Username(),
+		Host:     rep.Hostname(),
+		Path:     dirPath,
+	}
+
+	// Create a kopia snapshot
+	return SnapshotSource(ctx, rep, u, sourceInfo, root)
+}
+
+// OpenKopiaRepository connects to the kopia repository based on the config
+// NOTE: This assumes that `kopia repository connect` has been already run on the machine
+func OpenKopiaRepository(ctx context.Context, password string) (repo.Repository, error) {
+	if _, err := os.Stat(defaultConfigFileName()); os.IsNotExist(err) {
+		return nil, errors.New("Failed find kopia configuration file")
+	}
+
+	r, err := repo.Open(ctx, defaultConfigFileName(), password, &repo.Options{})
+	if os.IsNotExist(err) {
+		return nil, errors.New("Failed to find kopia repository, use `kopia repository connect`")
+	}
+
+	return r, errors.Wrap(err, "Failed to open kopia repository")
+}
+
+func defaultConfigFileName() string {
+	return filepath.Join(os.Getenv("HOME"), ".config", "kopia", "repository.config")
+}
+
+// SnapshotSource creates and uploads a kopia snapshot to the given repository
+func SnapshotSource(ctx context.Context, rep repo.Repository, u *snapshotfs.Uploader, sourceInfo snapshot.SourceInfo, rootDir fs.Entry) error {
+	fmt.Printf("Snapshotting %v ...\n", sourceInfo)
+
+	t0 := time.Now()
+
+	previous, err := findPreviousSnapshotManifest(ctx, rep, sourceInfo, nil)
+	if err != nil {
+		return errors.Wrap(err, "Failed to find previous kopia manifests")
+	}
+
+	policyTree, err := policy.TreeForSource(ctx, rep, sourceInfo)
+	if err != nil {
+		return errors.Wrap(err, "Failed to get kopia policy tree")
+	}
+
+	manifest, err := u.Upload(ctx, rootDir, policyTree, sourceInfo, previous...)
+	if err != nil {
+		return errors.Wrap(err, "Failed to upload the kopia snapshot")
+	}
+
+	manifest.Description = snapshotDescription
+
+	snapID, err := snapshot.SaveSnapshot(ctx, rep, manifest)
+	if err != nil {
+		return errors.Wrap(err, "Failed to save kopia manifest")
+	}
+
+	_, err = policy.ApplyRetentionPolicy(ctx, rep, sourceInfo, true)
```

Comment:
is this just a kopia-ism? Does this need to be called?

Reply:
Yes, we do need this if custom policy is setup for the repository. Default case has 2 snapshots for each frequency - 48, 14, 25 etc.